### PR TITLE
feat(common): add support for NGSI-LD tenants

### PR DESCRIPTION
**Pull request by bobeal**
- extract and check tenant in web filter (NGSILD-Tenant header)
- add r2dbc connection factory to dynamically point to the corresponding schema
- fine grained configuration of tenants (URI, KC realm issuer and DB schema)
- bind Flyway migrations for all known tenants
- return a NonexistentTenant error if providing an unknown tenant
- integrate tenant management in authentication layer
- add a `shared.properties` configuration file in `shared` module to avoid duplicating everything (as a consequence, also move `ApplicationProperties` data class in `shared`module)
- integrate tenant management in Kafka layer (producers and listeners)
- improve configuration of mock users in Web handlers tests

### CODE REVIEW

**Comment by bobeal** [comment]
@gpoujol there are still things to improve but let's go for a 1st review

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventService.kt [hunk 1]
```diff
@@ -200,11 +258,18 @@ class EntityEventService(
...
 
-private fun <A, B> Either<A, B>.logResults(eventsType: EventsType, entityId: URI) =
-    this.fold({
-        logger.error("Error while sending $eventsType event for entity $entityId: $it")
-    }, {
-        logger.debug("Successfully sent event {} event for entity {}", eventsType, entityId)
-    })
+    private fun <A, B> Either<A, B>.logEntityEvent(eventsType: EventsType, entityId: URI, tenantUri: URI) =
+        this.fold({
+            logger.error("Error while sending $eventsType event for entity $entityId: $it")
```
Comment (gpoujol) [inline search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventService.kt:264]: Maybe add `in tenant {}` also here

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventService.kt [hunk 2]
```diff
@@ -200,11 +258,18 @@ class EntityEventService(
...
+        }, {
+            logger.debug("Sent {} event for entity {} in tenant {}", eventsType, entityId, tenantUri)
+        })
+
+    private fun <A, B> Either<A, B>.logAttributeEvent(eventCategory: String, entityId: URI, tenantUri: URI) =
+        this.fold({
+            logger.error("Error while sending $eventCategory event for entity $entityId: $it")
```
Comment (gpoujol) [inline search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventService.kt:271]: Maybe add `in tenant {}` also here

File: search-service/src/test/kotlin/com/egm/stellio/search/web/AnonymousUserHandlerTests.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package com.egm.stellio.search.web
```
Comment (gpoujol) [inline search-service/src/test/kotlin/com/egm/stellio/search/web/AnonymousUserHandlerTests.kt:1]: why not put it directly in `EntityHandlerTest.kt` like it's done in `TemporalEntityOperationHandlerTest.kt` ?

File: shared/src/main/kotlin/com/egm/stellio/shared/model/ErrorResponse.kt [hunk 4]
```diff
@@ -108,5 +115,6 @@ enum class ErrorType(val type: URI) {
     ACCESS_DENIED(URI("https://uri.etsi.org/ngsi-ld/errors/AccessDenied")),
     NOT_IMPLEMENTED(URI("https://uri.etsi.org/ngsi-ld/errors/NotImplemented")),
     UNSUPPORTED_MEDIA_TYPE(URI("https://uri.etsi.org/ngsi-ld/errors/UnsupportedMediaType")),
-    NOT_ACCEPTABLE(URI("https://uri.etsi.org/ngsi-ld/errors/NotAcceptable"))
+    NOT_ACCEPTABLE(URI("https://uri.etsi.org/ngsi-ld/errors/NotAcceptable")),
+    NONEXISTENT_TENANT(URI("https://uri.etsi.org/ngsi-ld/errors/NonexistentTenant"))
```
Comment (gpoujol) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/ErrorResponse.kt:119]: NonExistentTenant*

Comment (gpoujol) [reply shared/src/main/kotlin/com/egm/stellio/shared/model/ErrorResponse.kt:119]: Sorry I saw it was in the spec. I just found it weird

File: shared/src/test/kotlin/com/egm/stellio/shared/web/TenantWebFilterTests.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Test
+    fun `it should return a NonexistentTenant error if the tenant does not exist`() {
+        webClient.get()
+            .uri("/router/mockkedroute/ok")
+            .header(NGSILD_TENANT_HEADER, "urn:ngsi-ld:tenant:02")
+            .exchange()
+            .expectStatus().isEqualTo(HttpStatus.NOT_FOUND)
+            .expectBody()
+            .jsonPath("$..type").isEqualTo("https://uri.etsi.org/ngsi-ld/errors/NonexistentTenant")
```
Comment (gpoujol) [inline shared/src/test/kotlin/com/egm/stellio/shared/web/TenantWebFilterTests.kt:58]: Maybe use directly `NONEXISTENT_TENANT`

**Review by gpoujol — APPROVED** [review]

File: search-service/src/test/kotlin/com/egm/stellio/search/web/AnonymousUserHandlerTests.kt (hunk 3 above)
Comment (bobeal) [reply search-service/src/test/kotlin/com/egm/stellio/search/web/AnonymousUserHandlerTests.kt:1]: I also removed it from `TemporalEntityOperationHandlerTest.kt`. kind of dumb to duplicate the same test in each and every handler for a transversal topic.

File: shared/src/main/kotlin/com/egm/stellio/shared/model/ErrorResponse.kt (hunk 4 above)
Comment (bobeal) [reply shared/src/main/kotlin/com/egm/stellio/shared/model/ErrorResponse.kt:119]: yes, in English, it is one word: https://dictionary.cambridge.org/dictionary/english/nonexistent

File: shared/src/test/kotlin/com/egm/stellio/shared/web/TenantWebFilterTests.kt (hunk 5 above)
Comment (bobeal) [reply shared/src/test/kotlin/com/egm/stellio/shared/web/TenantWebFilterTests.kt:58]: in such a test, I prefer testing with the real value the end-user will have in the response

**Pull request merged**